nntrainer/dataset/data_producers.h

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   data_producers.h
+ * @date   09 July 2021
+ * @brief  This file contains data producer interface
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#ifndef __DATA_PRODUCERS_H__
+#define __DATA_PRODUCERS_H__
+
+#include <functional>
+#include <limits>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <tensor.h>
+#include <tensor_dim.h>
+
+namespace nntrainer {
+
+/**
+ * @brief DataProducer interface used to abstract data provider
+ *
+ */
+class DataProducer {
+public:
+  using Iteration = std::tuple<bool, std::vector<Tensor>, std::vector<Tensor>>;
+
+  /**
+   * @brief create an iteration
+   * @return Iteration iteration, if std::get<0>(retval) == true means end of
+   * iteration, at the end of the iteration, it's responsibility of @a this to
+   * shuffle.
+   */
+  using Gernerator = std::function<Iteration(void)>;
+
+  constexpr inline static unsigned long long SIZE_UNDEFINED =
+    std::numeric_limits<unsigned long long>::max();
+
+  /**
+   * @brief Destroy the Data Loader object
+   *
+   */
+  virtual ~DataProducer() {}
+
+  /**
+   * @brief Get the layer type
+   * @return const std::string type representation
+   */
+  virtual const std::string getType() const = 0;
+
+  /**
+   * @brief Set the Property object
+   *
+   * @param properties properties to set
+   */
+  virtual void setProperty(const std::vector<std::string> &properties) {}
```

Comment:
how about making `setProperty()` either pure virtual or make the default implementation simply throw?

Reply:
I'll make this throw if property is not empty, this is similar to noop for data producers that are not using setProperty simply it does not neeed it.

Reply:
works.